### ISSUE TITLE
7.0rc2 java.lang.NullPointerException: input File:path/to/foo.pic.o.d belongs to a tree artifact whose metadata is missing

### ISSUE THREAD

**Issue by beasleyr-vmw** [bazelbuild/bazel#19988]
### Description of the bug:

While experimenting with Bazel 7.0rc2, I have run into multiple crashes with the following stack trace. These actions compile source code which is generated by another action during the build. The outputs of the code generator are unknown in advance, and so they are emitted into a tree artifact.

```
FATAL: bazel crashed due to an internal error. Printing stack trace:
java.lang.RuntimeException: Unrecoverable error while evaluating node 'ActionLookupData0{actionLookupKey=ActionTemplateExpansionKey{actionLookupKey=ConfiguredTargetKey{label=//mypackage:mylabel, config=BuildConfigurationKey[digest]}, actionIndex=0}, actionIndex=0}' (requested by nodes 'File:[[<execution_root>]bazel-out/k8-dbg-obj/bin]mypackage/_pic_objs/mylabel/mysubdir')
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:550)
        at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:414)
        at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinTask.doExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.scan(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinPool.runWorker(Unknown Source)
        at java.base/java.util.concurrent.ForkJoinWorkerThread.run(Unknown Source)
Caused by: java.lang.NullPointerException: input File:[[<execution_root>]bazel-out/k8-dbg-obj/bin]mypackage/_pic_dotd/mylabel/mysubdir/foo.pic.o.d belongs to a tree artifact whose metadata is missing
        at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:1010)
        at com.google.devtools.build.lib.remote.AbstractActionInputPrefetcher.maybeGetSymlink(AbstractActionInputPrefetcher.java:408)
        at com.google.devtools.build.lib.remote.AbstractActionInputPrefetcher.prefetchFile(AbstractActionInputPrefetcher.java:336)
        at com.google.devtools.build.lib.remote.AbstractActionInputPrefetcher.lambda$prefetchFiles$0(AbstractActionInputPrefetcher.java:299)
        at io.reactivex.rxjava3.internal.operators.flowable.FlowableFlatMapSingle$FlatMapSingleSubscriber.onNext(FlowableFlatMapSingle.java:118)
        at io.reactivex.rxjava3.internal.operators.flowable.FlowableFromIterable$IteratorSubscription.fastPath(FlowableFromIterable.java:185)
        at io.reactivex.rxjava3.internal.operators.flowable.FlowableFromIterable$BaseRangeSubscription.request(FlowableFromIterable.java:129)
        at io.reactivex.rxjava3.internal.operators.flowable.FlowableFlatMapSingle$FlatMapSingleSubscriber.onSubscribe(FlowableFlatMapSingle.java:106)
        at io.reactivex.rxjava3.internal.operators.flowable.FlowableFromIterable.subscribe(FlowableFromIterable.java:69)
        at io.reactivex.rxjava3.internal.operators.flowable.FlowableFromIterable.subscribeActual(FlowableFromIterable.java:47)
        at io.reactivex.rxjava3.core.Flowable.subscribe(Flowable.java:15917)
        at io.reactivex.rxjava3.internal.operators.flowable.FlowableFlatMapSingle.subscribeActual(FlowableFlatMapSingle.java:53)
        at io.reactivex.rxjava3.core.Flowable.subscribe(Flowable.java:15917)
        at io.reactivex.rxjava3.internal.operators.flowable.FlowableCollectSingle.subscribeActual(FlowableCollectSingle.java:53)
        at io.reactivex.rxjava3.core.Single.subscribe(Single.java:4855)
        at io.reactivex.rxjava3.internal.operators.single.SingleFlatMapCompletable.subscribeActual(SingleFlatMapCompletable.java:44)
        at io.reactivex.rxjava3.core.Completable.subscribe(Completable.java:2859)
        at io.reactivex.rxjava3.internal.operators.completable.CompletablePeek.subscribeActual(CompletablePeek.java:51)
        at io.reactivex.rxjava3.core.Completable.subscribe(Completable.java:2859)
        at com.google.devtools.build.lib.remote.util.RxFutures.toListenableFuture(RxFutures.java:195)
        at com.google.devtools.build.lib.remote.AbstractActionInputPrefetcher.prefetchFiles(AbstractActionInputPrefetcher.java:314)
        at com.google.devtools.build.lib.remote.AbstractActionInputPrefetcher.finalizeAction(AbstractActionInputPrefetcher.java:666)
        at com.google.devtools.build.lib.remote.RemoteOutputService.finalizeAction(RemoteOutputService.java:183)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.actuallyCompleteAction(SkyframeActionExecutor.java:1238)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.executeAction(SkyframeActionExecutor.java:1171)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.run(SkyframeActionExecutor.java:1062)
        at com.google.devtools.build.lib.skyframe.ActionExecutionState.runStateMachine(ActionExecutionState.java:165)
        at com.google.devtools.build.lib.skyframe.ActionExecutionState.getResultOrDependOnFuture(ActionExecutionState.java:94)
        at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:559)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:859)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.computeInternal(ActionExecutionFunction.java:333)
        at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:171)
        at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:461)
        ... 7 more
```

### Which category does this issue belong to?

Remote Execution

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

I don't yet know. I haven't been able to come up with a toy reproduction for this yet.

### Which operating system are you running Bazel on?

Linux

### Is this a regression? If yes, please try to identify the Bazel commit where the bug was introduced.

https://github.com/bazelbuild/bazel/commit/76fcd8d2d0734753adc18dbf5ecaf39f3b833755

Note: I couldn't use `bazelisk --bisect` this time because, while bisecting, some changes led to unrelated build breaks. I resorted to old school `git bisect` to determine hashes. Still, kudos to the folks at Google who've wired up the CI such that I can feed an arbitrary hash to bazelisk via `USE_BAZEL_VERSION=<hash>` and have things just work. :)

### Have you found anything relevant by searching the web?

No.

### Any other information, logs, or outputs that you want to share?

I have seen errors like this in the past when experimenting with BwoB, and so we've always avoided BwoB.  However, Bazel 7 is now enabling BwoB features by default, and so I'm using `--remote_download_outputs=all` in an attempt to explicitly disable the feature (including while bisecting).

It appears that I can work around this (though I'm not sure of the consequences) by adding `--noexperimental_inmemory_dotd_files` to my command line.

**Comment by coeuvre** [comment]
cc @tjgq

**Comment by coeuvre** [comment]
Thanks for the report and bisection! Working on the fix.

**Comment by beasleyr-vmw** [comment]
> A fix for this issue has been included in [Bazel 7.0.0 RC5](https://releases.bazel.build/7.0.0/rc5/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!

Confirmed resolution using RC5. Thanks!
